File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/organisms/hero/OrgHeroSummary.kt
package com.perrystreet.woof.designsystem.atomic.organisms.hero

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import com.perrystreet.woof.designsystem.atomic._tokens.spacing.SpacingRoles
import com.perrystreet.woof.designsystem.atomic.atoms.placeholder.AtomTextPlaceholder
import com.perrystreet.woof.designsystem.atomic.atoms.placeholder.atomPlaceholderShimmer
import com.perrystreet.woof.designsystem.atomic.atoms.placeholder.roles.TextPlaceholderRole
import com.perrystreet.woof.designsystem.atomic.atoms.text.AtomText
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextColorRole
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextFontRole
import com.perrystreet.woof.designsystem.atomic.molecules.header.MolTitleSubtitle
import com.perrystreet.woof.designsystem.atomic.molecules.placeholder.MolTagGroupPlaceholder
import com.perrystreet.woof.designsystem.atomic.molecules.tag.MolTagGroup

@Composable
fun OrgHeroSummary(
    title: String,
    subtitle: String?,
    tags: List<String>,
) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(SpacingRoles.Component.Cozy.dp),
    ) {
        when (subtitle) {
            null -> {
                OrgHeroSummaryTitleLoading(title = title)
                MolTagGroupPlaceholder()
            }
            else -> {
                MolTitleSubtitle(title = title, subtitle = subtitle)
                MolTagGroup(tags = tags)
            }
        }
    }
}

@Composable
private fun OrgHeroSummaryTitleLoading(title: String) {
    Column(verticalArrangement = Arrangement.spacedBy(SpacingRoles.Component.ExtraCompact.dp)) {
        AtomText(
            text = title,
            textFontRole = TextFontRole.DisplayH1,
            colorRole = TextColorRole.OnScrim,
            maxLines = 1,
        )
        Column(modifier = Modifier.atomPlaceholderShimmer()) {
            AtomTextPlaceholder(role = TextPlaceholderRole.Subtitle)
        }
    }
}

File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Organisms/Hero/OrgHeroSummary.swift
import SwiftUI

public struct OrgHeroSummary: View {
    private let title: String
    private let subtitle: String?
    private let tags: [String]

    public init(title: String, subtitle: String?, tags: [String]) {
        self.title = title
        self.subtitle = subtitle
        self.tags = tags
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: SpacingRoles.Component.cozy.rawValue) {
            switch subtitle {
            case .none:
                OrgHeroSummaryTitleLoading(title: title)
                MolTagGroupPlaceholder()
            case .some(let subtitle):
                MolTitleSubtitle(title: title, subtitle: subtitle)
                MolTagGroup(tags: tags)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

private struct OrgHeroSummaryTitleLoading: View {
    let title: String

    var body: some View {
        VStack(alignment: .leading, spacing: SpacingRoles.Component.extraCompact.rawValue) {
            AtomText(text: title, textFontRole: .displayH1, colorRole: .onScrim, maxLines: 1)
            AtomTextPlaceholder(role: .subtitle)
                .atomPlaceholderShimmer()
        }
    }
}
